### PR TITLE
Feat/enforce custom provider type

### LINKED ISSUE #24: Feat: Enforce custom provider types
# Description

We want to allow users to prefer one type of custom provider over the others. For instance, [factory providers over value providers can significantly increase performance](https://github.com/nestjs/nest/pull/12753).

# Acceptance Criteria

Given the options below:

```ts
// ...
rules: {
  '@trilon/enforce-custom-provider-type': [ 
   'warn', {  
     prefer: 'factory'
   }
  ]
}
```

Then, the following code will be considered **Invalid**:

```ts
const customValueProvider: Provider = {
  provide: 'TOKEN',
  useValue: 'some-value' // ⚠️ provider is not of type "factory"
}

const customClassProvider: Provider = {
  provide: AbstractClass,
  useClass: SomeClass  // ⚠️ provider is not of type "factory"
}
```

And the following code is considered **Valid**:

```ts
const factoryProvider: Provider = {
  provide: 'TOKEN',
  useFactory: () => 'some-value'
}
```

### PR DESCRIPTION
**Pull request by thiagomini**
Closes #24 

This is still a WIP. We need additional test cases and cover scenarios where we rename the `Provider` type

### CODE REVIEW

**Comment by tolgap** [comment]
@thiagomini I took the liberty to add a test case for where we alias `Provider` type as a starter to contribute. Hope you don't mind 🙂

**Comment by thiagomini** [comment]
That's awesome, @tolgap , your help is very much appreciated! Thanks for contributing to our cause 🙌

File: src/rules/enforce-custom-provider-type.rule.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+        specifiers
+          .filter(isImportSpecifier)
+          .filter(isProviderImport)
+          .forEach((spec) =>
+            providerTypesImported.push(spec.local.name ?? spec.imported.name)
+          );
+      },
+
+      'Identifier[typeAnnotation.typeAnnotation.type="TSTypeReference"]': (
```
Comment (tuxmachine) [inline src/rules/enforce-custom-provider-type.rule.ts:83]: Does this also capture providers inside `@Module({ providers: [] })`

Comment (thiagomini) [reply src/rules/enforce-custom-provider-type.rule.ts:83]: Good question Rick, I don't think so haha I'll have to create another test for that

Comment (thiagomini) [reply src/rules/enforce-custom-provider-type.rule.ts:103]: I believe this is now fixed, @tuxmachine , could you take a look in the tests and see if I'm missing something? thanks!

**Review by tolgap — COMMENTED** [review]
I think this is looking great @thiagomini 🔥 great work.

**Review by tolgap — APPROVED** [review]
LGTM! Great work @thiagomini . I do agree with the `engines.node` setting being too strict 😄. Would stop a lot of users of using this package.

**Pull request merged**